core/src/main/scala/kafka/controller/KafkaControllerZkUtils.scala

```
@@ -573,18 +574,86 @@ class KafkaControllerZkUtils(zookeeperClient: ZookeeperClient, isSecure: Boolean
+    retryUntilConnected(setDataRequest).asInstanceOf[SetDataResponse]
+  }
+
+  private def createConsumerOffset(group: String, topicPartition: TopicPartition, offset: Long): CreateResponse = {
+    val path = ConsumerOffset.path(group, topicPartition.topic, topicPartition.partition)
+    val createRequest = CreateRequest(path, ConsumerOffset.encode(offset), acls(path), CreateMode.PERSISTENT, null)
+    var createResponse = retryUntilConnected(createRequest).asInstanceOf[CreateResponse]
+    if (Code.get(createResponse.rc) == Code.NONODE) {
+      createRecursive(path.substring(0, path.lastIndexOf('/')))
+      createResponse = retryUntilConnected(createRequest).asInstanceOf[CreateResponse]
+    }
+    createResponse
+  }
+
   private def deleteRecursive(path: String): Unit = {
     val getChildrenResponse = retryUntilConnected(GetChildrenRequest(path, null)).asInstanceOf[GetChildrenResponse]
     if (Code.get(getChildrenResponse.rc) == Code.OK) {
       getChildrenResponse.children.foreach(child => deleteRecursive(s"$path/$child"))
-      val deleteResponse = retryUntilConnected(DeleteRequest(path, -1, null))
+      val deleteResponse = retryUntilConnected(DeleteRequest(path, Version.noVersion, null))
       if (Code.get(deleteResponse.rc) != Code.OK && Code.get(deleteResponse.rc) != Code.NONODE) {
         throw KeeperException.create(Code.get(deleteResponse.rc))
       }
     } else if (Code.get(getChildrenResponse.rc) != Code.NONODE) {
       throw KeeperException.create(Code.get(getChildrenResponse.rc))
     }
   }
+
+  private def createRecursive(path: String): Unit = {
+    val createRequest = CreateRequest(path, null, acls(path), CreateMode.PERSISTENT, null)
+    var createResponse = retryUntilConnected(createRequest).asInstanceOf[CreateResponse]
+    if (Code.get(createResponse.rc) == Code.NONODE) {
+      val parentPath = path.substring(0, path.lastIndexOf('/'))
```

We probably want to throw an exception if path.lastIndexOf('/') returns -1?